src/pybind/mgr/diskprediction/module.py

```diff
@@ -0,0 +1,346 @@
+        for _task in self._tasks:
+            if isinstance(_task, SmartTask):
+                msg = 'run smart agent successfully'
+                _task.event.set()
+        return 0, msg, ''
+
+    def _status(self, inbuf, cmd):
+        msg = 'diskprediction plugin status: %s' % self.status
+        return 0, msg, ''
+
+    def handle_command(self, inbuf, cmd):
+        for o_cmd in self.COMMANDS:
+            if cmd['prefix'] == o_cmd['cmd'][:len(cmd['prefix'])]:
+                fun = getattr(
+                    self, '_%s' % o_cmd['cmd'].split(' ')[1].replace('-', '_'))
+                if fun:
+                    return fun(inbuf, cmd)
+        return 0, 'cmd not found', ''
+
+    def show_module_config(self):
+        self.fsid = self.get('mon_map')['fsid']
+        self.log.debug('Found Ceph fsid %s', self.fsid)
+
+        for key, default in self.config_keys.items():
+            self.set_config_option(key, self.get_config(key, default))
+
+    def serve(self):
+        self.log.info('Starting diskprediction module')
+        self.run = True
+        self.status = DP_MGR_STAT_ENABLED
+
+        while self.run:
+            if self.get_configuration("diskprediction_config_mode").lower() in ['cloud', 'onpremise']:
+                enable_colud = True
+            if enable_colud and not self._activate_cloud:
+                self.start_cloud_disk_prediction()
+            elif not enable_colud and self._activate_cloud:
+                self.stop_cloud_disk_prediction()
+            self.event.wait(60)
+
+    def start_cloud_disk_prediction(self):
+        try:
+            if not self._activate_cloud:
+                for dp_task in DP_TASK:
+                    obj_task = dp_task(self)
+                    if obj_task:
+                        obj_task.run()
+                    else:
+                        raise Exception('failed to start task %s' % obj_task._task_name)
+                    self._tasks.append(obj_task)
+                self._activate_cloud = True
+                self.log.info('start cloud disk prediction')
+        except:
+            self.log.error('failed to start cloud disk prediction')
+            if self._tasks:
+                self._activate_cloud = True
+                self.stop_cloud_disk_prediction()
+
+    def stop_cloud_disk_prediction(self):
+        try:
+            if self._activate_cloud:
+                self.status = DP_MGR_STAT_DISABLED
+                while self._tasks:
+                    try:
+                        dp_task = self._tasks.pop()
+                        dp_task.terminate()
+                        del dp_task
+                    except:
+                        break
+                self._activate_cloud = False
+                self.log.info("stop cloud disk prediction")
+        except:
+            self.log.error('failed to stop cloud disk prediction')
+
+    def shutdown(self):
+        self.run = False
+        self.stop_cloud_disk_prediction()
```

Comment:
Instead of here, you can put the call to stop_cloud_disk_prediction at the end of serve(), so that it gets called after serve falls out of its loop.  That way you don't potentially have a start in serve() racing with a stop in shutdown().